File: scripts/cloudflare-deploy.js
#!/usr/bin/env node

/**
 * Standalone Cloudflare Pages Deployment Script
 * 
 * This script deploys static websites to Cloudflare Pages using the REST API.
 * No Wrangler CLI required!
 * 
 * Requirements:
 * - Node.js 14+
 * - Valid Cloudflare API token with Pages:Edit permissions
 * - Cloudflare Account ID
 * 
 * Usage:
 *   node cloudflare-deploy.js --account=abc123 --token=xxx --project=my-site --dir=./out
 * 
 * Environment Variables (alternative to CLI args):
 *   CLOUDFLARE_ACCOUNT_ID - Your Cloudflare Account ID
 *   CLOUDFLARE_API_TOKEN - Your Cloudflare API token
 *   CLOUDFLARE_PROJECT_NAME - Your Pages project name
 *   DEPLOY_DIR - Directory containing files to deploy (default: ./out)
 * 
 * Note: This script is designed for text-based files (HTML, CSS, JS).
 * For sites with binary assets (images, fonts), use Wrangler CLI or the web interface.
 */

const fs = require('fs').promises;
const path = require('path');
const https = require('https');

// Configuration
const config = {
  accountId: process.env.CLOUDFLARE_ACCOUNT_ID || getArg('--account'),
  apiToken: process.env.CLOUDFLARE_API_TOKEN || getArg('--token'),
  projectName: process.env.CLOUDFLARE_PROJECT_NAME || getArg('--project'),
  deployDir: process.env.DEPLOY_DIR || getArg('--dir') || './out',
  branch: process.env.DEPLOY_BRANCH || getArg('--branch') || 'main',
};

// Helper to get CLI arguments
function getArg(name) {
  const arg = process.argv.find(a => a.startsWith(name + '='));
  return arg ? arg.split('=')[1] : null;
}

// Helper to make HTTPS requests
function httpsRequest(url, options = {}) {
  return new Promise((resolve, reject) => {
    const urlObj = new URL(url);
    const reqOptions = {
      hostname: urlObj.hostname,
      path: urlObj.pathname + urlObj.search,
      method: options.method || 'GET',
      headers: {
        'Authorization': `Bearer ${config.apiToken}`,
        ...(options.headers || {}),
      },
    };

    console.log(`📊 DEBUG: Making ${reqOptions.method} request to Cloudflare API`);

    const req = https.request(reqOptions, (res) => {
      let data = '';
      res.on('data', chunk => data += chunk);
      res.on('end', () => {
        console.log(`📊 DEBUG: Response status: ${res.statusCode}`);
        
        if (res.statusCode >= 200 && res.statusCode < 300) {
          try {
            const parsed = JSON.parse(data);
            resolve({ status: res.statusCode, data: parsed });
          } catch (e) {
            resolve({ status: res.statusCode, data: data });
          }
        } else {
          console.error(`❌ HTTP ${res.statusCode} Error - ${reqOptions.method} to ${urlObj.pathname}`);
          reject(new Error(`HTTP ${res.statusCode}: ${res.statusText || 'Request failed'}`));
        }
      });
    });

    req.on('error', (error) => {
      console.error('❌ Request error:', error.message);
      reject(error);
    });

    if (options.body) {
      const bodyStr = typeof options.body === 'string' ? options.body : JSON.stringify(options.body);
      console.log(`📊 DEBUG: Request body size: ${bodyStr.length} bytes`);
      req.write(bodyStr);
    }

    req.end();
  });
}

// Check if project exists
async function checkProject() {
  console.log(`🔍 Checking if Cloudflare Pages project exists: ${config.projectName}`);
  try {
    await httpsRequest(
      `https://api.cloudflare.com/client/v4/accounts/${config.accountId}/pages/projects/${config.projectName}`
    );
    console.log('✅ Project exists');
    return true;
  } catch (error) {
    console.log('ℹ️  Project does not exist, will create it');
    return false;
  }
}

// Create project
async function createProject() {
  console.log(`📝 Creating Cloudflare Pages project: ${config.projectName}`);
  
  try {
    await httpsRequest(
      `https://api.cloudflare.com/client/v4/accounts/${config.accountId}/pages/projects`,
      {
        method: 'POST',
        headers: { 'Content-Type': 'application/json' },
        body: {
          name: config.projectName,
          production_branch: config.branch,
        },
      }
    );
    console.log('✅ Project created');
    return true;
  } catch (error) {
    console.error('❌ Failed to create project:', error.message);
    return false;
  }
}

// Read all files from directory recursively
async function readDirectory(dir, baseDir = dir) {
  const files = [];
  const items = await fs.readdir(dir, { withFileTypes: true });

  for (const item of items) {
    const fullPath = path.join(dir, item.name);
    
    if (item.isDirectory()) {
      files.push(...await readDirectory(fullPath, baseDir));
    } else if (item.isFile()) {
      try {
        const content = await fs.readFile(fullPath, 'utf-8');
        const relativePath = '/' + path.relative(baseDir, fullPath).replace(/\\/g, '/');
        
        // Validate file size (25MB limit per file for Cloudflare)
        const sizeInMB = Buffer.byteLength(content, 'utf-8') / 1024 / 1024;
        if (sizeInMB > 25) {
          console.warn(`⚠️  Warning: ${relativePath} is ${sizeInMB.toFixed(2)}MB (limit: 25MB), skipping...`);
          continue;
        }
        
        files.push({ path: relativePath, content });
      } catch (error) {
        console.warn(`⚠️  Warning: Failed to read ${fullPath}: ${error.message}`);
      }
    }
  }

  return files;
}

// Deploy to Cloudflare Pages
async function deployToCloudflare(files) {
  console.log(`🚀 Starting deployment to Cloudflare Pages...`);
  console.log(`📊 DEBUG: Deploying ${files.length} files`);
  console.log(`📊 DEBUG: Total size: ${files.reduce((sum, f) => sum + f.content.length, 0)} bytes`);

  // Create deployment
  console.log('📝 Creating deployment...');
  console.log(`📊 DEBUG: Branch: ${config.branch}, Stage: production`);
  
  const deployResponse = await httpsRequest(
    `https://api.cloudflare.com/client/v4/accounts/${config.accountId}/pages/projects/${config.projectName}/deployments`,
    {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: {
        branch: config.branch,
        stage: 'production', // Required: "production" or "preview"
      },
    }
  );

  console.log(`📊 DEBUG: Deploy response status: ${deployResponse.status}`);
  
  const uploadUrl = deployResponse.data.result?.upload_url;
  const deploymentId = deployResponse.data.result?.id;
  const stage = deployResponse.data.result?.stage;

  if (!uploadUrl) {
    console.error('❌ ERROR: No upload URL in response. Success:', deployResponse.data.success);
    throw new Error('No upload URL received from Cloudflare. Check API permissions and project settings.');
  }

  console.log(`✅ Deployment created (ID: ${deploymentId}, Stage: ${stage})`);
  console.log('📤 Uploading files...');

  // Create file manifest
  const manifest = {};
  for (const file of files) {
    const base64Content = Buffer.from(file.content, 'utf-8').toString('base64');
    manifest[file.path] = base64Content;
    console.log(`   📄 ${file.path} (${(file.content.length / 1024).toFixed(2)} KB → ${(base64Content.length / 1024).toFixed(2)} KB base64)`);
  }

  console.log(`📊 DEBUG: Total files in manifest: ${Object.keys(manifest).length}`);
  console.log(`📊 DEBUG: Total manifest size: ${JSON.stringify(manifest).length} bytes`);

  // Upload manifest
  console.log('📤 Uploading manifest to Cloudflare...');
  const uploadResponse = await httpsRequest(uploadUrl, {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: { manifest },
  });

  console.log(`📊 DEBUG: Upload response status: ${uploadResponse.status}`);
  console.log('✅ Files uploaded successfully');

  return {
    url: `https://${config.projectName}.pages.dev`,
    deploymentId,
  };
}

// Main deployment function
async function deploy() {
  console.log('\n☁️  Cloudflare Pages Deployment Tool\n');

  // Validate configuration
  console.log('\n🔧 Configuration:');
  console.log(`   Account ID: ${config.accountId ? 'configured' : 'NOT SET'}`);
  console.log(`   API Token: ${config.apiToken ? 'configured' : 'NOT SET'}`);
  console.log(`   Project Name: ${config.projectName || 'NOT SET'}`);
  console.log(`   Deploy Dir: ${config.deployDir}`);
  console.log(`   Branch: ${config.branch}`);
  console.log('');
  
  if (!config.accountId) {
    console.error('❌ Error: Cloudflare Account ID is required');
    console.error('Provide via --account=xxx or CLOUDFLARE_ACCOUNT_ID env var');
    console.error('\nFind your Account ID at: https://dash.cloudflare.com/');
    process.exit(1);
  }

  if (!config.apiToken) {
    console.error('❌ Error: Cloudflare API token is required');
    console.error('Provide via --token=xxx or CLOUDFLARE_API_TOKEN env var');
    console.error('\nCreate a token at: https://dash.cloudflare.com/profile/api-tokens');
    console.error('Required permission: Account → Cloudflare Pages → Edit');
    process.exit(1);
  }

  if (!config.projectName) {
    console.error('❌ Error: Project name is required');
    console.error('Provide via --project=my-site or CLOUDFLARE_PROJECT_NAME env var');
    process.exit(1);
  }

  try {
    // Step 1: Check/create project
    const projectExists = await checkProject();
    if (!projectExists) {
      const created = await createProject();
      if (!created) {
        process.exit(1);
      }
    }

    // Step 2: Read files
    console.log(`\n📂 Reading files from: ${config.deployDir}`);
    const files = await readDirectory(config.deployDir);
    console.log(`✅ Found ${files.length} file(s)`);

    if (files.length === 0) {
      console.error('❌ Error: No files found to deploy');
      process.exit(1);
    }

    // Step 3: Deploy
    const { url, deploymentId } = await deployToCloudflare(files);

    // Success!
    console.log('\n🎉 Deployment successful!');
    console.log(`\n🌐 Your site is live at: ${url}`);
    console.log(`📋 Deployment ID: ${deploymentId}\n`);

  } catch (error) {
    console.error('\n❌ Deployment failed:', error.message);
    process.exit(1);
  }
}

// Run deployment
if (require.main === module) {
  deploy();
}

module.exports = { deploy };

File: scripts/cloudflare-deploy.py
#!/usr/bin/env python3

"""
Standalone Cloudflare Pages Deployment Script

This script deploys static websites to Cloudflare Pages using the REST API.
No Wrangler CLI required!

Requirements:
- Python 3.7+
- requests library (pip install requests)
- Valid Cloudflare API token with Pages:Edit permissions
- Cloudflare Account ID

Usage:
    python3 cloudflare-deploy.py --account=abc123 --token=xxx --project=my-site --dir=./out

Environment Variables (alternative to CLI args):
    CLOUDFLARE_ACCOUNT_ID - Your Cloudflare Account ID
    CLOUDFLARE_API_TOKEN - Your Cloudflare API token
    CLOUDFLARE_PROJECT_NAME - Your Pages project name
    DEPLOY_DIR - Directory containing files to deploy (default: ./out)

Note: This script is designed for text-based files (HTML, CSS, JS).
For sites with binary assets (images, fonts), use Wrangler CLI or the web interface.
"""

import os
import sys
import argparse
import json
import base64
from pathlib import Path
from typing import List, Dict, Tuple

try:
    import requests
except ImportError:
    print("❌ Error: requests library is required")
    print("Install it with: pip install requests")
    sys.exit(1)


class CloudflareDeployer:
    """Cloudflare Pages deployment using REST API"""

    def __init__(self, account_id: str, api_token: str, project_name: str, deploy_dir: str, branch: str = "main"):
        self.account_id = account_id
        self.api_token = api_token
        self.project_name = project_name
        self.deploy_dir = Path(deploy_dir)
        self.branch = branch
        self.session = requests.Session()
        self.session.headers.update({
            "Authorization": f"Bearer {api_token}",
            "Content-Type": "application/json"
        })

    def check_project_exists(self) -> bool:
        """Check if Cloudflare Pages project exists"""
        print(f"🔍 Checking if Cloudflare Pages project exists: {self.project_name}")
        
        url = f"https://api.cloudflare.com/client/v4/accounts/{self.account_id}/pages/projects/{self.project_name}"
        response = self.session.get(url)
        
        if response.status_code == 200:
            print("✅ Project exists")
            return True
        else:
            print("ℹ️  Project does not exist, will create it")
            return False

    def create_project(self) -> bool:
        """Create Cloudflare Pages project"""
        print(f"📝 Creating Cloudflare Pages project: {self.project_name}")
        
        url = f"https://api.cloudflare.com/client/v4/accounts/{self.account_id}/pages/projects"
        payload = {
            "name": self.project_name,
            "production_branch": self.branch,
        }
        
        response = self.session.post(url, json=payload)
        
        if response.status_code in (200, 201):
            print("✅ Project created")
            return True
        else:
            print(f"❌ Failed to create project: {response.text}")
            return False

    def read_files(self) -> List[Dict[str, str]]:
        """Read all files from deploy directory"""
        print(f"\n📂 Reading files from: {self.deploy_dir}")
        
        if not self.deploy_dir.exists():
            raise FileNotFoundError(f"Deploy directory not found: {self.deploy_dir}")
        
        files = []
        MAX_FILE_SIZE_MB = 25  # Cloudflare limit
        
        for file_path in self.deploy_dir.rglob("*"):
            if file_path.is_file():
                try:
                    with open(file_path, "r", encoding="utf-8", errors="replace") as f:
                        content = f.read()
                    
                    # Validate file size
                    size_mb = len(content.encode("utf-8")) / 1024 / 1024
                    if size_mb > MAX_FILE_SIZE_MB:
                        print(f"⚠️  Warning: {file_path.name} is {size_mb:.2f}MB (limit: {MAX_FILE_SIZE_MB}MB), skipping...")
                        continue
                    
                    # Get relative path with leading slash
                    relative_path = "/" + str(file_path.relative_to(self.deploy_dir)).replace("\\", "/")
                    
                    files.append({
                        "path": relative_path,
                        "content": content
                    })
                except Exception as e:
                    print(f"⚠️  Warning: Failed to read {file_path.name}: {e}")
        
        print(f"✅ Found {len(files)} file(s)")
        return files

    def deploy_files(self, files: List[Dict[str, str]]) -> Tuple[str, str]:
        """Deploy files to Cloudflare Pages"""
        print(f"🚀 Starting deployment to Cloudflare Pages...")
        
        # Create deployment
        print("📝 Creating deployment...")
        
        url = f"https://api.cloudflare.com/client/v4/accounts/{self.account_id}/pages/projects/{self.project_name}/deployments"
        payload = {
            "branch": self.branch,
        }
        
        response = self.session.post(url, json=payload)
        
        if not response.ok:
            raise Exception(f"Failed to create deployment: {response.text}")
        
        deploy_data = response.json()
        upload_url = deploy_data.get("result", {}).get("upload_url")
        deployment_id = deploy_data.get("result", {}).get("id", "unknown")
        
        if not upload_url:
            raise Exception("No upload URL received from Cloudflare")
        
        print("✅ Deployment created, uploading files...")
        
        # Create file manifest
        manifest = {}
        for file in files:
            # Encode content to base64
            content_bytes = file["content"].encode("utf-8")
            base64_content = base64.b64encode(content_bytes).decode("utf-8")
            manifest[file["path"]] = base64_content
            
            size_kb = len(file["content"]) / 1024
            print(f"   - {file['path']} ({size_kb:.2f} KB)")
        
        # Upload manifest
        upload_response = self.session.post(upload_url, json={"manifest": manifest})
        
        if not upload_response.ok:
            raise Exception(f"Failed to upload files: {upload_response.text}")
        
        print("✅ Files uploaded successfully")
        
        # Construct deployment URL
        deployment_url = f"https://{self.project_name}.pages.dev"
        
        return deployment_url, deployment_id

    def deploy(self) -> str:
        """Main deployment workflow"""
        print("\n☁️  Cloudflare Pages Deployment Tool (Python)\n")
        
        # Step 1: Check/create project
        project_exists = self.check_project_exists()
        if not project_exists:
            created = self.create_project()
            if not created:
                raise Exception("Failed to create project")
        
        # Step 2: Read files
        files = self.read_files()
        if not files:
            raise Exception("No files found to deploy")
        
        # Step 3: Deploy
        deployment_url, deployment_id = self.deploy_files(files)
        
        return deployment_url, deployment_id


def get_config() -> Tuple[str, str, str, str, str]:
    """Get configuration from CLI args or environment variables"""
    parser = argparse.ArgumentParser(
        description="Deploy static website to Cloudflare Pages using REST API"
    )
    parser.add_argument("--account", help="Cloudflare Account ID")
    parser.add_argument("--token", help="Cloudflare API token")
    parser.add_argument("--project", help="Cloudflare Pages project name")
    parser.add_argument("--dir", default="./out", help="Directory to deploy (default: ./out)")
    parser.add_argument("--branch", default="main", help="Deployment branch (default: main)")
    
    args = parser.parse_args()
    
    account_id = args.account or os.getenv("CLOUDFLARE_ACCOUNT_ID")
    api_token = args.token or os.getenv("CLOUDFLARE_API_TOKEN")
    project_name = args.project or os.getenv("CLOUDFLARE_PROJECT_NAME")
    deploy_dir = args.dir or os.getenv("DEPLOY_DIR", "./out")
    branch = args.branch or os.getenv("DEPLOY_BRANCH", "main")
    
    return account_id, api_token, project_name, deploy_dir, branch


def main():
    """Main entry point"""
    try:
        account_id, api_token, project_name, deploy_dir, branch = get_config()
        
        # Validate configuration
        if not account_id:
            print("❌ Error: Cloudflare Account ID is required")
            print("Provide via --account=xxx or CLOUDFLARE_ACCOUNT_ID env var")
            print("\nFind your Account ID at: https://dash.cloudflare.com/")
            sys.exit(1)
        
        if not api_token:
            print("❌ Error: Cloudflare API token is required")
            print("Provide via --token=xxx or CLOUDFLARE_API_TOKEN env var")
            print("\nCreate a token at: https://dash.cloudflare.com/profile/api-tokens")
            print("Required permission: Account → Cloudflare Pages → Edit")
            sys.exit(1)
        
        if not project_name:
            print("❌ Error: Project name is required")
            print("Provide via --project=my-site or CLOUDFLARE_PROJECT_NAME env var")
            sys.exit(1)
        
        # Create deployer and deploy
        deployer = CloudflareDeployer(account_id, api_token, project_name, deploy_dir, branch)
        deployment_url, deployment_id = deployer.deploy()
        
        # Success!
        print("\n🎉 Deployment successful!")
        print(f"\n🌐 Your site is live at: {deployment_url}")
        print(f"📋 Deployment ID: {deployment_id}\n")
        
    except Exception as e:
        print(f"\n❌ Deployment failed: {e}")
        sys.exit(1)


if __name__ == "__main__":
    main()
